04-onion-routing.md

```diff
@@ -150,52 +155,75 @@ The overall structure of the packet is as follows:
 2. data:
    * [`1`:`version`]
    * [`33`:`public_key`]
-   * [`20*65`:`hops_data`]
+   * [`20*FRAME_SIZE`:`hops_data`]
    * [`32`:`hmac`]
 
 For this specification (_version 0_), `version` has a constant value of `0x00`.
 
-The `hops_data` field is a structure that holds obfuscations of the next hop's
-address, transfer information, and its associated HMAC. It is 1300 bytes (`20x65`) long
-and has the following structure:
+The `hops_data` field is list of `hop_payload`s.
+A `hop_payload` is a structure that holds obfuscations of the next hop's address, transfer information, and its associated HMAC.
+Each `hop_payload` consists of one of more `frame`s to hold the information that the hop should receive.
+The total number of `frame`s used by all `hop_payload`s MUST NOT exceed 20, but MAY be less than that, in which case the remaining frames are padded to reach the full 1300 byte length for the `hops_data`.
+
+The `hops_data` has the following structure (`n1` and `n2` being the number of frames used by hops `1` and `2` respectively):
 
 1. type: `hops_data`
 2. data:
-   * [`1`:`realm`]
-   * [`32`:`per_hop`]
-   * [`32`:`HMAC`]
+   * [`n1*FRAME_SIZE`: `hop_payload`]
+   * [`n2*FRAME_SIZE`: `hop_payload`]
    * ...
    * `filler`
 
-Where, the `realm`, `per_hop` (with contents dependent on `realm`), and `HMAC`
-are repeated for each hop; and where, `filler` consists of obfuscated,
-deterministically-generated padding, as detailed in
-[Filler Generation](#filler-generation). Additionally, `hops_data` is
-incrementally obfuscated at each hop.
+Where `filler` consists of obfuscated, deterministically-generated padding, as detailed in [Filler Generation](#filler-generation).
+Additionally, `hop_payload`s are incrementally obfuscated at each hop.
 
-The `realm` byte determines the format of the `per_hop` field; currently, only `realm`
-0 is defined, for which the `per_hop` format follows:
+Each `hop_payload` has the following structure:
 
-1. type: `per_hop` (for `realm` 0)
+1. type: `hop_payload`
+2. data:
+   * [`1`: `num_frames_and_realm`]
```

Comment:
What happened to the packet type? IMO we still shouldn't intermingle them, as it's feasible that a distinct realm may also warrant an entirely different interpretation of the TLV values.